File: src/main/java/dev/retrotv/crypto/owe/kdf/pbkdf2/Pbkdf2.java
package dev.retrotv.crypto.owe.kdf.pbkdf2;

import dev.retrotv.crypto.owe.kdf.KDF;
import dev.retrotv.utils.PasswordStrengthUtil;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.crypto.password.Pbkdf2PasswordEncoder;

/**
 * Pbkdf2 알고리즘으로 암호화 하기 위한 {@link KDF} 추상 클래스의 구현체 입니다.
 * Spring Security의 {@link PasswordEncoder}와 호환됩니다.
 * @author  yjj8353
 * @since   1.8
 */
public class Pbkdf2 extends KDF {
    private final Pbkdf2PasswordEncoder pbkdf2PasswordEncoder;

    public Pbkdf2() {
        pbkdf2PasswordEncoder = Pbkdf2PasswordEncoder.defaultsForSpringSecurity_v5_8();
    }

    public Pbkdf2(CharSequence secret, int saltLength, int iterations,
           Pbkdf2PasswordEncoder.SecretKeyFactoryAlgorithm secretKeyFactoryAlgorithm) {
        if (secret == null) {
            throw new IllegalArgumentException("secret은 null일 수 없습니다.");
        }

        if (secretKeyFactoryAlgorithm == null) {
            throw new IllegalArgumentException("secretKeyFactoryAlgorithm은 null일 수 없습니다.");
        }

        pbkdf2PasswordEncoder = new Pbkdf2PasswordEncoder(secret, saltLength, iterations, secretKeyFactoryAlgorithm);
    }

    @Override
    public String encode(CharSequence rawPassword) {
        if (rawPassword == null) {
            throw new IllegalArgumentException("rawPassword는 null일 수 없습니다.");
        }

        return pbkdf2PasswordEncoder.encode(rawPassword);
    }

    @Override
    public boolean matches(CharSequence rawPassword, String encodedPassword) {
        if (rawPassword == null || encodedPassword == null) {
            return false;
        }

        return pbkdf2PasswordEncoder.matches(rawPassword, encodedPassword);
    }

    @Override
    public boolean upgradeEncoding(String encodedPassword) {
        if (encodedPassword == null) {
            return false;
        }

        return PasswordStrengthUtil.checkLength(8, encodedPassword) &&
               PasswordStrengthUtil.isInclude(
                   true,
                   false,
                   false,
                   true,
                   true,
                   encodedPassword
               );
	}
}

File: src/main/kotlin/dev/retrotv/crypto/owe/password/pbkdf2/Pbkdf2.kt
package dev.retrotv.crypto.owe.password.pbkdf2

import dev.retrotv.crypto.owe.password.KDF
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.crypto.password.Pbkdf2PasswordEncoder
import org.springframework.security.crypto.password.Pbkdf2PasswordEncoder.SecretKeyFactoryAlgorithm

/**
 * Pbkdf2 알고리즘으로 암호화 하기 위한 [KDF] 추상 클래스의 구현체 입니다.
 * Spring Security의 [PasswordEncoder]와 호환됩니다.
 * @author  yjj8353
 * @since   1.0.0
 */
class Pbkdf2 : KDF {
    private val pbkdf2PasswordEncoder: Pbkdf2PasswordEncoder

    constructor() {
        pbkdf2PasswordEncoder = Pbkdf2PasswordEncoder.defaultsForSpringSecurity_v5_8()
    }

    constructor(
        secret: CharSequence?, saltLength: Int, iterations: Int,
        secretKeyFactoryAlgorithm: SecretKeyFactoryAlgorithm?
    ) {
        pbkdf2PasswordEncoder = Pbkdf2PasswordEncoder(secret, saltLength, iterations, secretKeyFactoryAlgorithm)
    }

    override fun encode(rawPassword: CharSequence): String {
        return pbkdf2PasswordEncoder.encode(rawPassword)
    }

    override fun matches(rawPassword: CharSequence, encodedPassword: String?): Boolean {
        return if (encodedPassword == null) {
            false
        } else pbkdf2PasswordEncoder.matches(rawPassword, encodedPassword)
    }
}
